src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs

```diff
@@ -0,0 +1,2255 @@
+                );
+        }
+
+        [Fact]
+        public void TestWithAccessModifiers_Protected_WithoutGetter()
+        {
+            string source = @"
+public class C
+{
+    public string Property { protected init { throw null; } }
+}
+";
+            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (4,19): error CS0276: 'C.Property': accessibility modifiers on accessors may only be used if the property or indexer has both a get and a set accessor
+                //     public string Property { protected init { throw null; } }
+                Diagnostic(ErrorCode.ERR_AccessModMissingAccessor, "Property").WithArguments("C.Property").WithLocation(4, 19)
+                );
+        }
+
+        [Fact]
+        public void OverrideScenarioWithSubstitutions()
+        {
+            string source = @"
+public class C<T>
+{
+    public string Property { get; init; }
+}
+public class Derived : C<string>
+{
+    void M()
+    {
+        Property = null; // 1
+    }
+
+    Derived()
+    {
+        Property = null;
+    }
+}
+";
+            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (10,9): error CS8802: Init-only member 'C<string>.Property' can only be assigned from a constructor, object initialization or 'with' expression of that type.
+                //         Property = null; // 1
+                Diagnostic(ErrorCode.ERR_AssignmentInitOnly, "Property").WithArguments("C<string>.Property").WithLocation(10, 9)
+                );
+        }
+
+        [Fact]
+        public void ImplementationScenarioWithSubstitutions()
+        {
+            string source = @"
+public interface I<T>
+{
+    public string Property { get; init; }
+}
+public class CWithInit : I<string>
+{
+    public string Property { get; init; }
+}
+public class CWithoutInit : I<string> // 1
+{
+    public string Property { get; set; }
+}
+";
+            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics(
+                // (10,29): error CS8804: 'CWithoutInit' does not implement interface member 'I<string>.Property.set'. 'CWithoutInit.Property.set' cannot implement 'I<string>.Property.set' because it does not match by init-only.
+                // public class CWithoutInit : I<string> // 1
+                Diagnostic(ErrorCode.ERR_CloseUnimplementedInterfaceMemberWrongInitOnly, "I<string>").WithArguments("CWithoutInit", "I<string>.Property.set", "CWithoutInit.Property.set").WithLocation(10, 29)
+                );
+        }
+
+        [Fact]
+        public void InLambdaOrLocalFunction()
+        {
+            string source = @"
+public class C<T>
```

Comment:
>C<T> [](start = 13, length = 4)

Consider testing similar scenarios in constructor of this type and inside the init setter.